File: packages/types/src/tool/approval.ts
/**
 * Tool Approval Types
 * Types for tool approval coordination
 */

import type { LLMToolCall } from "../message/index.js";
import type { AutoApprovalCategory } from "./risk-level.js";
import type { FilePermissionSettings } from "./file-permission.js";
import type { McpApprovalSettings, McpRequest } from "./mcp-approval.js";
import type { ToolExecutionResult } from "./execution.js";

/**
 * Security Preset
 * Predefined security configurations for tool approval
 */
export type SecurityPreset = "SAFE" | "BALANCED" | "PERMISSIVE";

/**
 * Workspace boundary control settings
 */
export interface WorkspaceBoundarySettings {
  /** Allow read operations outside workspace */
  allowReadOnlyOutsideWorkspace?: boolean;
  /** Allow write operations outside workspace */
  allowWriteOutsideWorkspace?: boolean;
}

/**
 * Command execution settings
 */
export interface CommandExecutionSettings {
  /** Allowed command prefixes (longest prefix match) */
  allowedCommands?: string[];
  /** Denied command prefixes (longest prefix match) */
  deniedCommands?: string[];
}

/**
 * Network request settings (for HTTP tools like web_fetch)
 */
export interface NetworkSettings {
  /** Allowed domains */
  allowedDomains?: string[];
  /** Denied domains */
  deniedDomains?: string[];
}

/**
 * Interaction settings
 */
export interface InteractionSettings {
  /** Followup question timeout (ms) for auto-selecting suggestion */
  followupAutoApproveTimeoutMs?: number;
}

/**
 * Tool Approval Options
 * Options for tool approval coordinator
 * Note: This is different from ToolApprovalConfig in workflow/config.ts
 * which is used for workflow-level configuration
 */
export interface ToolApprovalOptions {
  // === Main switch ===
  /** Enable auto-approval system (default: false) */
  autoApprovalEnabled?: boolean;

  // === Presets ===
  /** Security preset to apply (overrides manual category settings if provided) */
  securityPreset?: SecurityPreset;

  // === File permission (HIGHEST PRIORITY) ===
  /** File permission settings - evaluated first before any other checks */
  filePermissions?: FilePermissionSettings;

  // === Category-based settings ===
  /** Category-based auto-approval settings */
  categories?: Partial<Record<AutoApprovalCategory, boolean>>;

  // === Boundary settings ===
  /** Workspace boundary controls */
  workspaceBoundary?: WorkspaceBoundarySettings;
  /** Allow writing to protected files */
  allowWriteProtected?: boolean;

  // === Operation-specific settings ===
  /** Command execution settings */
  command?: CommandExecutionSettings;
  /** Network request settings (for HTTP tools) */
  network?: NetworkSettings;
  /** MCP approval settings */
  mcp?: McpApprovalSettings;
  /** Interaction settings */
  interaction?: InteractionSettings;

  // === Usage Limits ===
  /** Maximum number of consecutive auto-approved requests before requiring manual approval */
  maxAutoApprovedRequests?: number;

  // === General settings ===
  /** Approval timeout in milliseconds (0 = no timeout) */
  approvalTimeout?: number;
}

/**
 * Tool Approval Request
 * Request for tool approval
 */
export interface ToolApprovalRequest {
  /** Tool call to approve */
  toolCall: LLMToolCall;
  /** Tool description */
  toolDescription?: string;
  /** Context ID (execution ID, session ID, etc.) */
  contextId: string;
  /** Node ID (optional) */
  nodeId?: string;
  /** Interaction ID for tracking */
  interactionId: string;
  
  // Batch execution metadata
  /** Unique ID for this batch */
  batchId?: string;
  /** Position in batch (0-based) */
  toolIndex?: number;
  /** Total tools in batch */
  totalTools?: number;
  /** Tools remaining after this one */
  pendingQueue?: LLMToolCall[];
  /** Results from auto-executed prefix */
  autoExecutedResults?: ToolExecutionResult[];
}

/**
 * Tool Approval Result
 * Result of tool approval process
 */
export interface ToolApprovalResult {
  /** Whether the tool call is approved */
  approved: boolean;
  /** Tool call ID */
  toolCallId: string;
  /** Edited parameters (if user modified them) */
  editedParameters?: Record<string, unknown>;
  /** User instruction (optional additional context) */
  userInstruction?: string;
  /** User's comment explaining decision */
  annotation?: string;
  /** Rejection reason (if not approved) */
  rejectionReason?: string;
  
  // Batch control
  /** Should continue with remaining tools? */
  continueBatch?: boolean;
}

/**
 * Tool Approval Handler
 * Interface for handling tool approval requests
 */
export interface ToolApprovalHandler {
  /**
   * Request tool approval
   * @param request Approval request
   * @returns Approval result
   */
  requestApproval(request: ToolApprovalRequest): Promise<ToolApprovalResult>;
}

/**
 * Tool Batch Result
 * Result of processing a batch of tools
 */
export interface ToolBatchResult {
  /** Unique ID for this batch */
  batchId: string;
  /** Auto-executed tools */
  autoExecuted: ToolExecutionResult[];
  /** First tool needing approval */
  confirmationRequired: LLMToolCall | null;
  /** User's decision */
  confirmationResult?: ToolApprovalResult;
  /** Tools not yet processed */
  remainingQueue: LLMToolCall[];
  /** True if all tools done */
  allCompleted: boolean;
}

/**
 * Pending Tool Call
 * Information about a pending tool call for UI display
 */
export interface PendingToolCall {
  /** Tool call ID */
  id: string;
  /** Tool name */
  name: string;
  /** Tool arguments */
  arguments?: string;
  /** Risk level */
  riskLevel?: import("./risk-level.js").ToolRiskLevel;
}

/**
 * Tool Approval Coordinator Params
 * Parameters for tool approval coordinator
 */
export interface ToolApprovalCoordinatorParams {
  /** Tool call to process */
  toolCall: LLMToolCall;
  /** Approval options */
  options?: ToolApprovalOptions;
  /** Context ID */
  contextId: string;
  /** Node ID (optional) */
  nodeId?: string;
  /** Tool description (optional) */
  toolDescription?: string;
  /** Approval handler */
  approvalHandler: ToolApprovalHandler;
}

File: crates/wf-types/src/tool/approval.rs
use serde::{Deserialize, Serialize};

use super::file_permission::FilePermissionSettings;
use super::mcp_approval::McpApprovalSettings;
use super::ToolCall;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum SecurityPreset {
    Safe,
    Balanced,
    Permissive,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ApprovalCategories {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_read_only: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_write: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_execute: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_mcp: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_network: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_allow_interaction: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkspaceBoundary {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allow_read_only_outside_workspace: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allow_write_outside_workspace: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CommandApprovalSettings {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allowed_commands: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub denied_commands: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct NetworkApprovalSettings {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allowed_domains: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub denied_domains: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InteractionApprovalSettings {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub followup_auto_approve_timeout_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolApprovalOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_approval_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security_preset: Option<SecurityPreset>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub risk_threshold: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_approve_patterns: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub categories: Option<ApprovalCategories>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_boundary: Option<WorkspaceBoundary>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_permissions: Option<FilePermissionSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<CommandApprovalSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp: Option<McpApprovalSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub network: Option<NetworkApprovalSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interaction: Option<InteractionApprovalSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allow_write_protected: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolApprovalRequest {
    pub tool_call: ToolCall,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_description: Option<String>,
    pub context_id: super::super::Id,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub node_id: Option<String>,
    pub interaction_id: super::super::Id,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolApprovalResult {
    pub approved: bool,
    pub tool_call_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub edited_parameters: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rejection_reason: Option<String>,
}
